kubernetes-client-api/src/main/java/io/fabric8/kubernetes/client/dsl/ItemReplacable.java

```diff
@@ -16,14 +16,26 @@
 
 package io.fabric8.kubernetes.client.dsl;
 
-public interface Cascading<T> {
+public interface ItemReplacable<T> {
+
+  /**
+   * Similar to calling resource(item).replaceStatus()
```

Comment:
Should we aim to remove functionality duplication? More specifically, can we make it so that there's only one way to perform an operation? It would reduce the cognitive load of users.

Reply:
Yes that is the eventual goal.  There are several concerns.  The first is that it will be a major api change, which I'm now leaning towards using deprecations initially.  Based upon the changes in this pr it will be relatively easy to introduce CollectionItemReplacable and CollectionItemWritableOperation which are deprecated - so that calling configMaps().create(x) would be deprecated and the user would be pointed to configMaps().resource(x).create().

The second is that there are other points of inconsistency in the dsl.  patchStatus(item), as mentioned above, does not currently perform a diff, so it could be offered as patchStatus().  ProjectOperation.createProjectAndRoleBindings(name, ...) - does not belong as a collection level method, and instead should be projects().withName(name).createProjectAndRoleBindings(...).  Before we fully remove things we should probably identify what all of these are.

The last is that to be done completely will be a lot of work.  The core issue is the differentiation between calling configMaps().withName("x") and calling configMaps().resource(x) - they both return the same Resource, or Resource subclass.  However you shouldn't be able to call configMaps().withName("x").create(), nor does configMaps().resource(x).create(y) make sense.  In theory you'd like to split the Resource class, such that when withName is called you'd logically have:
```
NamedResource<T> extends
    FromServerGettable<T>,
    Lockable<? extends Deletable>,
    WatchAndWaitable<T>, Versionable<WatchAndWaitable<T>>,
    Deletable,
    DryRunable<? extends Deleteable>,
    Requirable<T>, Readiable, Informable<T>
```

Lockable and DryRunable would need to be converted to non-generic, but other than that it's representative.  Then

```
Resource<T> extends NamedResource<T>
    Lockable<ReplaceDeletable<T>>,
    WritableOperation<T>,
    DryRunable<WritableOperation<T>>
```

However the ability to extend a Resource - (PodResource, ServiceInstanceResource, etc.) is based only off of extending Resource.  You would instead need both NamedPodResource and PodResource.  So something like we currently have as:

```
public interface ServiceInstanceResource extends Resource<ServiceInstance> {

    ServiceBinding bind(String secretName);
}
```

You would actually want to be a mixin that is applied to both the NamedResource and the Resource - 
serviceInstances().withName("x").bind("secret") or serviceInstances().resource(x).bind("secret")

Alternatively you could view resource(x) as a way to mixin Lockable<ReplaceDeletable<T>>, WritableOperation<T>, DryRunable<WritableOperation<T>>

While there are relatively few Resource subclasses left, the code is setup to only know about one Resource type.  So the simplest solution for now is to just allow configMaps().withName("x").create() and similar methods to exist, but throw exceptions.